test/test-vcvars.sh

```diff
@@ -0,0 +1,127 @@
+# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+. "${0%/*}/test.sh"
+
+host=x64
+if [ "$(uname -m)" = "aarch64" ]; then
+    host=arm64
+fi
+
+BASE=$(. "${BIN}msvcenv.sh" && echo $BASE)
+ARCH=$(. "${BIN}msvcenv.sh" && echo $ARCH)
+
+if [ "$host" = "$ARCH" ]; then
+    vcvars_arch=${ARCH}
+else
+    vcvars_arch=${host}_${ARCH}
+fi
+
+cat >test-$vcvars_arch.bat <<EOF
+@echo off
+
+set "CWD=%CD%"
+
+set VSCMD_START_DIR=
+call $BASE\VC\Auxiliary\Build\vcvarsall.bat $vcvars_arch
+if %errorlevel% neq 0 exit /B %errorlevel%
+
+if not "%CWD%"=="%CD%" (
+    echo ERROR: vcvarsall.bat changed CWD to %CD%.
+    exit /B 2
+)
+
+set "WindowsSDKVersion=%WindowsSDKVersion:\=%"
+
+call :SaveUnixPath VSINSTALLDIR
+call :SaveUnixPath VCToolsInstallDir
+call :SaveUnixPath WindowsSdkDir
+call :SaveVariable WindowsSDKVersion
+call :SaveUnixPath UniversalCRTSdkDir
+call :SaveVariable UCRTVersion
+call :SearchInPath cl.exe
+call :SearchInPath rc.exe
+call :SearchInPath MSBuild.exe
+exit /B 0
+
+:SearchInPath
+setlocal EnableDelayedExpansion
+set "f=%1"
+set "f=%f:.=_%"
+for /F "delims=" %%G in ('where %1') do (
+    set "%f%=%%G"
+    call :SaveUnixPath %f%
+    exit /B 0
+)
+exit /B 1
+
+:SaveUnixPath
+setlocal EnableDelayedExpansion
+set "p=!%1!"
+if "%p:~-1%"=="\" set "p=%p:~0,-1%"
+for /F "delims=" %%G in ('winepath -u "%p%"') do (
+    set "%1=%%G"
+    call :SaveVariable %1
+)
+exit /B 0
+
+:SaveVariable
+setlocal EnableDelayedExpansion
+echo %1="!%1!" >> $vcvars_arch-env.txt
+exit /B 0
+EOF
+
+
+TestVariable() {
+    if [ "${!1}" != "$2" ]; then
+        echo "ERROR: $1=\"${!1}\""
+        return 1
+    fi
+    return 0
+}
+
+TestUnixPath() {
+    if [ "$(readlink -f "${!1}")" != "$(readlink -f "$2")" ]; then
+        echo "ERROR: $1=\"${!1}\""
+        return 1
+    fi
+    return 0
+}
+
+EXEC "" WINEDEBUG=-all $(command -v wine64 || command -v wine) cmd /c test-$vcvars_arch.bat || EXIT
+tr -d '\r' <$vcvars_arch-env.txt >$vcvars_arch-env
+. $vcvars_arch-env
+
+SDKBASE=$(. "${BIN}msvcenv.sh" && echo $SDKBASE)
+SDKBASE=${SDKBASE//\\//}
+SDKBASE=${SDKBASE#z:}
+
+MSVCDIR=$(. "${BIN}msvcenv.sh" && echo $MSVCDIR)
+MSVCDIR=${MSVCDIR//\\//}
+MSVCDIR=${MSVCDIR#z:}
+
+EXEC "" TestUnixPath VSINSTALLDIR       $(. "${BIN}msvcenv.sh" && echo $BASE_UNIX)
+EXEC "" TestUnixPath VCToolsInstallDir  $MSVCDIR
+EXEC "" TestUnixPath WindowsSdkDir      $SDKBASE
+EXEC "" TestVariable WindowsSDKVersion  $(. "${BIN}msvcenv.sh" && echo $SDKVER)
+EXEC "" TestUnixPath UniversalCRTSdkDir $SDKBASE
+EXEC "" TestUnixPath UCRTVersion        $(. "${BIN}msvcenv.sh" && echo $SDKVER)
+# Below tests require where.exe (available in Wine 9.3+).
+# EXEC "" TestUnixPath cl_exe             $(. "${BIN}msvcenv.sh" && echo $BINDIR)/cl.exe
+# EXEC "" TestUnixPath rc_exe             $(. "${BIN}msvcenv.sh" && echo $SDKBINDIR)/rc.exe
+# EXEC "" TestUnixPath MSBuild_exe        $(. "${BIN}msvcenv.sh" && echo $MSBUILDBINDIR)/MSBuild.exe
+
```

Comment:
Should conditionally enable these tests based on wine version, as Wine 10 is available on macOS.